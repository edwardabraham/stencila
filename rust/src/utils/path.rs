use lexiclean::Lexiclean;
use std::path::{Path, PathBuf};

/// Merge paths
///
/// Differs from `path.join()` in that it joins the *parent* of `path1`, not `path1` itself.
/// Does not canonicalize the path, or check for its existence.
pub fn merge<P1: AsRef<Path>, P2: AsRef<Path>>(path1: P1, path2: P2) -> PathBuf {
    path1.as_ref().join("..").join(path2).lexiclean()
}

#[cfg(test)]
mod test {
    use super::*;
    use path_slash::PathBufExt;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_merge() {
        assert_eq!(merge("a", "c").to_slash_lossy(), "c");
        assert_eq!(merge("a/b", "c").to_slash_lossy(), "a/c");
        assert_eq!(merge("a/b/../d/e", "c").to_slash_lossy(), "a/d/c")
    }
}
