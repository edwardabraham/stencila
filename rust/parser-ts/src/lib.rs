use once_cell::sync::Lazy;
use parser_treesitter::{
    eyre::Result,
    formats::Format,
    graph_triples::{relations, resources},
    parse_info, ParseInfo, Parser, ParserTrait, TreesitterParser,
};
use std::path::Path;

/// Tree-sitter based parser for JavaScript
static PARSER_JS: Lazy<TreesitterParser> = Lazy::new(|| {
    TreesitterParser::new(
        tree_sitter_typescript::language_typescript(),
        parser_js::QUERY,
    )
});

/// Tree-sitter based parser for TypeScript
static PARSER_TS: Lazy<TreesitterParser> =
    Lazy::new(|| TreesitterParser::new(tree_sitter_typescript::language_typescript(), QUERY));

/// Tree-sitter AST query for TypeScript
///
/// These are query patterns that extend those for JavaScript defined
/// in `parser-js`.
const QUERY: &str = r#"
(program
    [
        (variable_declaration
            (variable_declarator
                name: (identifier) @name
                type: (type_annotation) @type
            )
        )
        (lexical_declaration
            (variable_declarator
                name: (identifier) @name
                type: (type_annotation) @type
            )
        )
        (export_statement
            declaration: (lexical_declaration
                (variable_declarator
                    name: (identifier) @name
                    type: (type_annotation) @type
                )
            )
        )
    ]
)
"#;

/// A parser for TypeScript
pub struct TsParser {}

impl ParserTrait for TsParser {
    fn spec() -> Parser {
        Parser {
            language: Format::TypeScript.spec().title,
        }
    }

    fn parse(path: &Path, code: &str) -> Result<ParseInfo> {
        let code = code.as_bytes();
        let tree = PARSER_TS.parse(code);

        // Query the tree for typed patterns defined in this module
        let matches = PARSER_TS.query(code, &tree);
        let relations_typed = matches
            .iter()
            .filter_map(|(pattern, captures)| match pattern {
                0 => {
                    // Assigns a symbol at the top level of the module
                    let range = captures[0].range;
                    let name = captures[0].text.clone();
                    let type_annotation = captures[1].node;
                    let type_string = type_annotation
                        .named_child(0)
                        .and_then(|node| node.utf8_text(code).ok())
                        .unwrap_or_default();
                    let kind = match type_string {
                        "boolean" => "Boolean",
                        "number" => "Number",
                        "string" => "String",
                        "object" => "Object",
                        _ => {
                            if type_string.starts_with("Array<") {
                                "Array"
                            } else if type_string.starts_with("Record<string") {
                                "Object"
                            } else {
                                ""
                            }
                        }
                    };
                    Some((
                        relations::assigns(range),
                        resources::symbol(path, &name, kind),
                    ))
                }
                _ => None,
            });

        // Query the tree for untyped patterns defined in the JavaScript module
        let matches = PARSER_JS.query(code, &tree);
        let relations_untyped = matches.iter().filter_map(|(pattern, capture)| {
            parser_js::handle_patterns(path, code, pattern, capture)
        });

        let relations = relations_typed.chain(relations_untyped).collect();

        let parse_info = parse_info(path, &Self::spec().language, code, matches, 0, relations);
        Ok(parse_info)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_snaps::{insta::assert_json_snapshot, snapshot_fixtures};
    use test_utils::fixtures;

    #[test]
    fn parse_ts_fragments() {
        snapshot_fixtures("fragments/ts/*.ts", |path| {
            let code = std::fs::read_to_string(path).expect("Unable to read");
            let path = path.strip_prefix(fixtures()).expect("Unable to strip");
            let parse_info = TsParser::parse(path, &code).expect("Unable to parse");
            assert_json_snapshot!(parse_info);
        })
    }

    #[test]
    fn parse_js_fragments() {
        snapshot_fixtures("fragments/js/*.js", |path| {
            let code = std::fs::read_to_string(path).expect("Unable to read");
            let path = path.strip_prefix(fixtures()).expect("Unable to strip");
            let parse_info = TsParser::parse(path, &code).expect("Unable to parse");
            assert_json_snapshot!(parse_info);
        })
    }
}
