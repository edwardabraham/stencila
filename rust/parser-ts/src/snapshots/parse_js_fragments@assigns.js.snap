---
source: rust/parser-ts/src/lib.rs
assertion_line: 136
expression: resource_info
input_file: fixtures/fragments/js/assigns.js

---
{
  "resource": {
    "type": "Code",
    "path": "fragments/js/assigns.js",
    "id": "",
    "kind": "SoftwareSourceCode",
    "language": "JavaScript"
  },
  "relations": [
    [
      {
        "type": "Assign",
        "range": [
          0,
          0,
          0,
          4
        ]
      },
      {
        "type": "Symbol",
        "path": "fragments/js/assigns.js",
        "name": "var1",
        "kind": "Number"
      }
    ]
  ],
  "compile_digest": "ijXqJ2GHy-rL0gSvAY3aRJM4OWaBCA6UbwzWiNamOzs.vBja-GhNGn5S813yV_DNjJKYZ2_vpSXQiPwFaZ2fpvA..0.0"
}
